src/coreclr/jit/forwardsub.cpp

```diff
@@ -0,0 +1,392 @@
+        JITDUMP(" pinned local\n");
+        return false;
+    }
+
+    // Only fwd sub if we expect no code duplication
+    // We expect one def and one use.
+    //
+    if (varDsc->lvRefCnt(RCS_EARLY) != 2)
+    {
+        JITDUMP(" not asg (single-use lcl)\n");
+        return false;
+    }
+
+    // And local is unalised
+    //
+    if (varDsc->IsAddressExposed())
+    {
+        JITDUMP(" not asg (unaliased single-use lcl)\n");
+        return false;
+    }
+
+    // Could handle this case --perhaps-- but we'd want to update ref counts.
+    //
+    if (lvaIsImplicitByRefLocal(lclNum))
+    {
+        JITDUMP(" implicit by-ref local\n");
+        return false;
+    }
+
+    // Local seems suitable. See if the next statement
+    // contains the one and only use.
+    //
+    Statement* const nextStmt = stmt->GetNextStmt();
+
+    // We often see stale flags, eg call flags after inlining.
+    // Try and clean these up.
+    //
+    gtUpdateStmtSideEffects(nextStmt);
+    gtUpdateStmtSideEffects(stmt);
+
+    // Scan for the (single) use.
+    //
+    ForwardSubVisitor fsv(this, lclNum);
+    fsv.WalkTree(nextStmt->GetRootNodePointer(), nullptr);
+
+    assert(fsv.GetCount() <= 1);
+
+    if ((fsv.GetCount() == 0) || (fsv.GetNode() == nullptr))
+    {
+        JITDUMP(" no next stmt use\n");
+        return false;
+    }
+
+    JITDUMP(" [%06u] is only use of [%06u] (V%02u) ", dspTreeID(fsv.GetNode()), dspTreeID(lhsNode), lclNum);
+
+    // Next statement seems suitable.
+    // See if we can forward sub without changing semantics.
+    //
+    // We could just extract the value portion and forward sub that,
+    // but cleanup would be more complicated.
+    //
+    GenTree* const rhsNode      = rootNode->gtGetOp2();
+    GenTree* const nextRootNode = nextStmt->GetRootNode();
+    GenTree*       fwdSubNode   = rhsNode;
+
+    // Can't substitute a qmark (unless the use is RHS of an assign... could check for this)
+    // Can't substitute GT_CATCH_ARG.
+    //
+    if (fwdSubNode->OperIs(GT_QMARK, GT_CATCH_ARG))
+    {
+        JITDUMP(" node to sub is qmark or catch arg\n");
+        return false;
+    }
+
+    // Bail if sub node has embedded assignment.
+    //
+    if ((fwdSubNode->gtFlags & GTF_ASG) != 0)
+    {
+        JITDUMP(" node to sub has effects\n");
+        return false;
+    }
+
+    // Bail if types disagree.
+    //
+    if (fsv.GetNode()->TypeGet() != fwdSubNode->TypeGet())
+    {
+        JITDUMP(" mismatched types\n");
+        return false;
+    }
+
+    // We can forward sub if
+    //
+    // the value of the fwdSubNode can't change and its evaluation won't cause side effects,
+    //
+    // or,
+    //
+    // if the next tree can't change the value of fwdSubNode or be adversly impacted by fwdSubNode effects
+    //
+    const bool fwdSubNodeInvariant   = ((fwdSubNode->gtFlags & GTF_ALL_EFFECT) == 0);
+    const bool nextTreeIsPureUpToUse = ((fsv.GetFlags() & (GTF_EXCEPT | GTF_GLOB_REF | GTF_CALL)) == 0);
```

Comment:
It feels very risky to rely on `GTF_GLOB_REF` before (or, really, after, since we don't have the checker enabled for it) morph. For example:

```cs
static int* _lclAddr;

[MethodImpl(MethodImplOptions.NoInlining)]
static int Problem(int a, int b)
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    static void CaptureLcl(int* lclAddr)
    {
        _lclAddr = lclAddr;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static int MutateLcl()
    {
        (*_lclAddr)++;
        return 0;
    }

    CaptureLcl(&a);

    b = a + a;

    return MutateLcl() + b; // We will forward-substitute 'a' because "a + a" won't have GTF_GLOB_REF on it.
}
```

Reply:
Thanks for the example. Seems like we'll need to walk over the tree to substitute and compute the actual set of effects flags.

Reply:
Added some extra flags scanning.